2005/chuuuul/ArchitectureStudy/app/src/main/java/com/example/architecture/data/source/local/NaverLocalDataSourceImpl.kt

```diff
@@ -0,0 +1,66 @@
+package com.example.architecture.data.source.local
+
+import android.content.Context
+import com.example.architecture.data.model.MovieModel
+import com.example.architecture.util.ConstValue.Companion.MOVIE_SHARED_PREFERENCE_NAME
+import com.google.gson.Gson
+import com.google.gson.reflect.TypeToken
+
+class NaverLocalDataSourceImpl : NaverLocalDataSource {
+
+    private fun movieModelToString(movies: List<MovieModel>): String {
+        return Gson().toJson(movies)
+    }
+
+    private fun stringToMovieModel(gsonString: String): List<MovieModel> {
+        val collectionType = object : TypeToken<List<MovieModel>>() {}.type
+
+        return Gson().fromJson(gsonString, collectionType)
+    }
+
+    override fun getMovieList(
+        context: Context,
+        keyword: String,
+        onSuccess: (movies: List<MovieModel>) -> Unit
+    ) {
+        val sharedPref =
```

Comment:
각 메서드마다 SharedPreference 객체를 각각 계속 생성해주셨는데 이렇게 하신 이유가 context를 '메서드 호출시에만 부를 수 있어서'라면 NaverLocalDataSourceImpl (context: Context)형태로 받아서 property로 선언하면 중복코드를 없앨 수 있을 것 같은데 어떻게 생각하시나요?
제가 예측한것이 아니라면 혹시 각 객체를 따로따로 생성하시는 이유가 무엇인지 궁금합니다.

두번째로는 각 객체 생성시  ?: return으로 null처리를 해주셨습니다만 null인 경우는 어떠한 경우인건지 궁금합니다.

Reply:
> 두번째로는 각 객체 생성시 ?: return으로 null처리를 해주셨습니다만 null인 경우는 어떠한 경우인건지 궁금합니다.

저도 자세하게 모르겠지만
`context.getSharedPreferences(MOVIE_SHARED_PREFERENCE_NAME, Context.MODE_PRIVATE)`
의 반환값은 SharedPreferences! 가 됩니다. 즉 null 처리가 필요하다고 판단되었고, 공식문서에서도 null처리가 포함된 예제가 있었습니다.
 추측하건데, Context가 유효하지 않을 경우 null이 발생하지 않을까 싶습니다 !

Reply:
> 
말씀해주신 지적사항을 적용하려고 하니깐 Object로 선언한 클래스에는 생성자를 받을 수 없어 context를 local 클래스에 생성자로 넣어 줄 수가 없었습니다. 그래서 액티비티에 종속되는 Class를 사용하기로 하고 말씀해주신 대로 context를 프로퍼티로 사용하였습니다. Model을 Activity마다 갖는 경우 말고 싱글턴으로 가져가는 경우는 없는지 궁금하네요 ㅎㅎ
>

Reply:
> 감사합니다. 수정했습니다.
> context를 계속 받은 이유는 원래 Repository를 object로 생성하여 싱글턴으로 운용하려고 했습니다. Object 클래스로 만들어서 remote와 local의 data source를 가지고 있으면서 액티비티에 종속되지 않은 Model 객체를 생성하려고 했습니다. 그래서 액티비티에서 context를 받지 않고 필요할때마다 context를 매번 불러주어 사용하러고 했습니다.
> 말씀해주신 지적사항을 적용하려고 하니깐 Object로 선언한 클래스에는 생성자를 받을 수 없어 context를 local 클래스에 생성자로 넣어 줄 수가 없었습니다. 그래서 액티비티에 종속되는 Class를 사용하기로 하고 말씀해주신 대로 context를 프로퍼티로 사용하였습니다. Model을 Activity마다 갖는 경우 말고 싱글턴으로 가져가는 경우는 없는지 궁금하네요 ㅎㅎ

Reply:
c98e4c1
감사합니다. companion object를 써서 한번 구현해봤는데 어떤가요?

Reply:
[코틀린 컨벤션 참고]
레이아웃
일반적으로 클래스의 내용은 다음 순서로 정렬됩니다.

Property declarations and initializer blocks
Secondary constructors
Method declarations
Companion object
Do not sort the method declarations alphabetically or by visibility, and do not separate regular methods from extension methods. Instead, put related stuff together, so that someone reading the class from top to bottom can follow the logic of what's happening. Choose an order (either higher-level stuff first, or vice versa) and stick to it.

코틀린 컨벤션에 따라 바꿔주시면 좋을 것 같습니다.

Reply:
d064ad8, c7ae081
감사합니다. 수정했습니다. 또 하나 배워갑니다!